### PR TITLE
[new release] bam (2 packages) (0.1)

### PR DESCRIPTION
**Pull request by francoisthire**
A property-based testing library with internal shrinking

- Project page: <a href="https://github.com/francoisthire/bam">https://github.com/francoisthire/bam</a>
- Documentation: <a href="https://francoisthire.github.io/bam/">https://francoisthire.github.io/bam/</a>

##### CHANGES:

- Initial release of the Bam library.

### CODE REVIEW

File: packages/bam/bam.0.1/opam [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+license: "MIT"
+tags: ["test" "pbt" "shrinking" "internal"]
+homepage: "https://github.com/francoisthire/bam"
+doc: "https://francoisthire.github.io/bam/"
+bug-reports: "https://github.com/francoisthire/bam/issues"
+depends: [
+  "ocaml"
+  "dune" {>= "3.7"}
+  "pringo"
+  "zarith"
```
Comment (mseri) [inline packages/bam/bam.0.1/opam:16]:
```suggestion
  "zarith" {>= "1.9"}
```
I think this is where `Z.Compare` was added

File: packages/bam/bam.0.1/opam [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+doc: "https://francoisthire.github.io/bam/"
+bug-reports: "https://github.com/francoisthire/bam/issues"
+depends: [
+  "ocaml" {>= "4.14"}
+  "dune" {>= "3.7"}
+  "pringo"
+  "zarith" {>= "1.9"}
+  "odoc" {with-doc}
+  "tezt" {with-test}
+]
```
Comment (mseri) [inline packages/bam/bam.0.1/opam:19]:
```suggestion
  "tezt-bam {with-test & post}
]
```

File: packages/bam/bam.0.1/opam [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+license: "MIT"
+tags: ["test" "pbt" "shrinking" "internal"]
+homepage: "https://github.com/francoisthire/bam"
+doc: "https://francoisthire.github.io/bam/"
+bug-reports: "https://github.com/francoisthire/bam/issues"
+depends: [
+  "ocaml" {>= "4.14"}
+  "dune" {>= "3.7"}
+  "pringo"
+  "zarith" {>= "1.9"}
```
Comment (mseri) [inline packages/bam/bam.0.1/opam:16]:
```suggestion
  "zarith" {>= "1.13"}
```
for `Z.random_int_gen`

File: packages/bam/bam.0.1/opam (hunk 2 above)
Comment (mseri) [reply packages/bam/bam.0.1/opam:19]: Currently bam tests require tezt-bam

File: packages/bam/bam.0.1/opam [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+doc: "https://francoisthire.github.io/bam/"
+bug-reports: "https://github.com/francoisthire/bam/issues"
+depends: [
+  "ocaml" {>= "4.14"}
+  "dune" {>= "3.7"}
+  "pringo"
+  "zarith" {>= "1.13"}
+  "odoc" {with-doc}
+  "tezt" {with-test}
+  "tezt-bam {with-test & post}
```
Comment (mseri) [inline packages/bam/bam.0.1/opam:19]:
```suggestion
  "tezt-bam" {with-test & post}
```

**Comment by mseri** [comment]
Tests are currently broken due to circular dependencies, I would suggest to move them to tezt-bam if they require both libraries

**Comment by francoisthire** [comment]
Many thanks for your review and your patches.

- I have added version dependencies as you suggested
- I have tried to remove a dependency in the test library so that they are not run by `dune @runtest` when building `bam`. However, I am not sure it will work.

Maybe I could use some help here. 

The current repository has the following structure:

```bash
.
├── lib_bam
├── lib_tezt_bam
├── test
```

Is there a native way in `dune` to say that tests in `test` folder should be run only with `tezt-bam`?

**Comment by mseri** [comment]
In the dune file test stanzas you can specify the package. For example
```
(test
 (name mytest)
 (libraries alcotest tezt-bam uri)
 (package tezt-bam))
```
will make `mytest` run with the libraries specified there as dependencies when running the tests for the `tezt-bam` package

**Comment by francoisthire** [comment]
Just tried that, and I still observe some failure. While executing `dune @runtest` it seems he still want to get `tezt-bam` from the `test/dune` file. I don't understand why.

**Comment by francoisthire** [comment]
Small ping @mseri , Any suggestion to make this PR moving forward? Adding `(package tezt-bam)` seems not enough.

**Comment by mseri** [comment]
I think it is not using the correct sources, it is failing with wrong checksum:
```
2024-02-12 11:46.54: Error in tezt-bam.0.1:              error 60: Upstream check failed: "The archive doesn't match checksums:
                - archive: sha256=cd845d70aab1b728f41d3cf538b7a069a4981d0878eacf7ba4aca15025743e22, in opam file: sha256=50b5ee12c393b8d18cde20674bd3a5ec534f5ed52c31554c966b95f2f319b1a1
                - archive: sha512=8c0ca8b3aba5ab15c583f3f9d7cfaf2ce7869652cd1908f65f7ff2f8189dc0fada4ba04da10e10ea4a3b750761a3a6f9e915a133b6d346ad7c062e37197c5e3c, in opam file: sha512=e79c3b2b694558365ac448ae87521c21a9672874f63e81fcc4b23dc2a6fdf04f381c440770746069b7a2719b0a97a85a1969e164ee6adcccacb6a10ef07d3a13
              ."
2024-02-12 11:46.54: Error in tezt-bam.0.1: Failed to download the archive. Details: https://github.com/francoisthire/bam/releases/download/v0.1/bam-0.1.tbz (Bad checksum, expected sha256=50b5ee12c393b8d18cde20674bd3a5ec534f5ed52c31554c966b95f2f319b1a1)
2024-02-12 11:46.54: Error in bam.0.1:              error 60: Upstream check failed: "The archive doesn't match checksums:
                - archive: sha256=cd845d70aab1b728f41d3cf538b7a069a4981d0878eacf7ba4aca15025743e22, in opam file: sha256=50b5ee12c393b8d18cde20674bd3a5ec534f5ed52c31554c966b95f2f319b1a1
                - archive: sha512=8c0ca8b3aba5ab15c583f3f9d7cfaf2ce7869652cd1908f65f7ff2f8189dc0fada4ba04da10e10ea4a3b750761a3a6f9e915a133b6d346ad7c062e37197c5e3c, in opam file: sha512=e79c3b2b694558365ac448ae87521c21a9672874f63e81fcc4b23dc2a6fdf04f381c440770746069b7a2719b0a97a85a1969e164ee6adcccacb6a10ef07d3a13
              ."
2024-02-12 11:46.54: Error in bam.0.1: Failed to download the archive. Details: https://github.com/francoisthire/bam/releases/download/v0.1/bam-0.1.tbz (Bad checksum, expected sha256=50b5ee12c393b8d18cde20674bd3a5ec534f5ed52c31554c966b95f2f319b1a1)
2024-02-12 11:48.43: Job failed: 4 errors
```

**Comment by francoisthire** [comment]
Should be better now. Let me know if I can do something about the remaining failures.

**Comment by mseri** [comment]
Looks good now :)

**Pull request merged**